enhancements/multi-arch/heterogeneous-architecture.md

```diff
@@ -0,0 +1,348 @@
+payloads which CVO will need to look at to make upgrade recommendations. There are a few upgrade scenarios to consider with the addition of
+heterogeneous payloads:
+
+- heterogeneous -> heterogeneous payload upgrade
+- homogeneous -> heterogeneous payload upgrade
+- heterogeneous -> homogeneous payload upgrade
+
+The support for homogeneous payloads will exist so customers running disconnected homogeneous clusters do not incur the cost of mirroring the
+additional architectures if they were to use heterogeneous payloads. This means that in order for CVO to find the current release in the graph, it
+will have to know about the existing homogeneous graph as well in addition to the new heterogeneous graph. There are few options on the best way
+to make update recommendations especially in the case of upgrading from a homogeneous to a heterogeneous payload:
+
+- Have the heterogeneous graph include the homogeneous releases so CVO can make recommendations based on the current homogeneous release image.
+This is complicated by the fact that there will be multiple releases with the same version but different payloads. The trigger to update to a
+heterogeneous payload would be an option provided by the administrator through either the ClusterVersion or Infrastructure API to "move" to the
+heterogeneous graph.
+- Have both the graphs be completely independent of each other and have an option in one of the aforementioned APIs to make the cluster heterogeneous
+and CVO would request the heterogeneous graph, find the current release version and update to it.
+- Have an independent channel for heterogeneous payloads and have administrators switch channels if they want to move between homogeneous and
+heterogeneous payloads
+
+These options are under discussion and a finalized proposal will be put forth in the ticket for upgrades referenced above along with the API that would
+change to include the option to indicate a move to the heterogeneous payload.
+
+Cincinnati needs to [support][OCPPLAN-7643] parsing manifestlist images to be able to upgrade to a heterogeneous architecture payload.
+
+### Imagestreams
+
+Imagestreams need to better support manifestlist images. As of today they do recognize manifestlists, but it only imports images matching the control
+plane architecture. The idea here would be to import the manifest listed image in its entirety. This also has other implications such as the fact that
+the internal registry does not support manifest listed images. So, with full manifestlist support in imagestreams, pullthrough is not possible for
+manifestlists. We would ask users to use quay.io rather than use the internal registry if full manifestlist support is needed.
+
+The oc commands like debug, must-gather etc use the imagestreams and therefore would use an x86 image to deploy on the worker nodes which would fail
```

Comment:
Can we just change these image streams (must-gather, etc) to use `reference: true`? I don't think we really need image stream functionality (ability to trigger new deployments, etc.) for these images.

```
spec:
  tags:
  - from:
      kind: DockerImage
      name: registry...@sha256:...
    name: latest
    reference: true
```

Reply:
remind me, does that effectively mean the imagestreamtag SHA will just be the manifestlist SHA?

Reply:
from.name will be copied as is into status:

```
status:
  tags:
  - items:
    - created: "2022-02-02T19:01:17Z"
      dockerImageReference: registry...@sha256:...
      generation: 3
      image: ""
    tag: latest
```

With such a change:
```
$ oc get -n openshift istag must-gather:latest
Error from server (NotFound): imagestreamtags.image.openshift.io "must-gather:latest" not found
$ oc get -n openshift imagetag must-gather:latest
NAME                 SPEC        STATUS                   HISTORY   UPDATED
must-gather:latest   Scheduled   registry...@sha256:...   1         3 hours ago
```

Reply:
i guess since from.name is already a SHA for those in-payload imagestreams, it works ok

Reply:
i just checked that adding the `reference: true`  tag to the tools imagestream works fine when i do an oc debug node on a hetero arch cluster for any of the nodes. so this is something that would have to be added in the imagestream manifests in cluster samples right? and they would add it only for a heterogeneous deployment? would it cause complications during upgrade to homogeneous ? or would we never use the internal registry for these images going forward?

Reply:
it will make impossible to use pull-through for these images, i.e. you won't be able to pull it as `image-registry.openshift-image-registry.svc:5000/openshift/must-gather:latest`

Reply:
CVO may need to learn about this flag: https://github.com/openshift/cluster-version-operator/blob/d3f4ca856badfe3db7a74415bd4a418d128999b1/lib/resourcemerge/imagestream.go#L32